component/http/v2/component.go

```diff
@@ -0,0 +1,95 @@
+// Package v2 provides a ready to use HTTP component.
+package v2
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"net/http"
+	"sync"
+	"time"
+
+	"github.com/beatlabs/patron/log"
+)
+
+const (
+	port                = 50000
+	readTimeout         = 30 * time.Second
+	writeTimeout        = 60 * time.Second
+	idleTimeout         = 240 * time.Second
+	shutdownGracePeriod = 5 * time.Second
+)
+
+// Component implementation of an HTTP router.
+type Component struct {
+	port                int
+	readTimeout         time.Duration
+	writeTimeout        time.Duration
+	shutdownGracePeriod time.Duration
+	handler             http.Handler
+	sync.Mutex
+	certFile string
+	keyFile  string
+}
+
+func New(handler http.Handler, oo ...OptionFunc) (*Component, error) {
+	if handler == nil {
+		return nil, errors.New("handler is nil")
+	}
+
+	cmp := &Component{
+		port:                port,
+		readTimeout:         readTimeout,
+		writeTimeout:        writeTimeout,
+		shutdownGracePeriod: shutdownGracePeriod,
+		handler:             handler,
+	}
+
+	for _, option := range oo {
+		err := option(cmp)
+		if err != nil {
+			return nil, err
+		}
+	}
+
+	return cmp, nil
+}
+
+// Run starts the HTTP server.
+func (c *Component) Run(ctx context.Context) error {
+	c.Lock()
+	chFail := make(chan error)
+	srv := c.createHTTPServer()
+	go c.listenAndServe(srv, chFail)
+	c.Unlock()
+
+	select {
+	case <-ctx.Done():
+		log.Info("shutting down HTTP component")
+		ctx, cancel := context.WithTimeout(context.Background(), c.shutdownGracePeriod)
+		defer cancel()
+		return srv.Shutdown(ctx)
+	case err := <-chFail:
+		return err
+	}
+}
+
+func (c *Component) createHTTPServer() *http.Server {
+	return &http.Server{
+		Addr:         fmt.Sprintf(":%d", c.port),
+		ReadTimeout:  c.readTimeout,
+		WriteTimeout: c.writeTimeout,
```

Comment:
I'm not a big fan of `WriteTimeout`. First, its behavior depends on whether TLS is enabled, making it more complex to understand and use. Furthermore, it closes the TCP connection without returning a proper HTTP code if the timeout is reached. Meanwhile, it doesn't propagate the cancellation to the handler context. Therefore, a handler may continue its execution without knowing that the TCP connection is already closed.

Since, Go 1.8, `http.TimeoutHandler` has been released to overcome these limitations:

```go 
&http.Server{
	// ...
	Handler:           http.TimeoutHandler(handler, time.Second, "foo"),
}
```

Reply:
Nice catch, working on it!

Reply:
In order to be consistent we need to define a new functional option for this timeouthandler and setup a default value.